src/os/bluestore/BlueStoreSlowFastCoDel.cc

```diff
@@ -0,0 +1,274 @@
+
+  if (activated && max_queue_length < get_kv_throttle_current()) {
+    max_queue_length = get_kv_throttle_current();
+  }
+  if (min_latency == INITIAL_LATENCY_VALUE || latency < min_latency) {
+    min_latency = latency;
+  }
+  slow_interval_txc_cnt++;
+  slow_interval_registered_bytes += txc_bytes;
+}
+
+void BlueStoreSlowFastCoDel::on_min_latency_violation() {
+  if (target_latency > 0) {
+    double diff = (double) (target_latency - min_latency);
+    auto error_ratio = std::abs(diff) / min_latency;
+    if (error_ratio > 0.5) {
+      error_ratio = 0.5;
+    }
+    bluestore_budget = std::max(bluestore_budget * (1 - error_ratio),
+                                min_bluestore_budget * 1.0);
+  }
+}
+
+void BlueStoreSlowFastCoDel::on_no_violation() {
+  if (bluestore_budget < max_queue_length * 1.5) {
+    bluestore_budget = bluestore_budget + bluestore_budget_increment;
+  }
+}
+
+void BlueStoreSlowFastCoDel::on_config_changed(CephContext *cct) {
+  {
+    std::lock_guard l(register_lock);
+
+    activated = cct->_conf->bluestore_codel;
+    target_slope = cct->_conf->bluestore_codel_throughput_latency_tradeoff;
+    slow_interval = (int64_t) millisec_to_nanosec(
+            cct->_conf->bluestore_codel_slow_interval);
+    initial_fast_interval = (int64_t) millisec_to_nanosec(
+            cct->_conf->bluestore_codel_fast_interval);
+    initial_target_latency = (int64_t) millisec_to_nanosec(
+            cct->_conf->bluestore_codel_initial_target_latency);
+    min_target_latency = (int64_t) millisec_to_nanosec(
+            cct->_conf->bluestore_codel_min_target_latency);
+    max_target_latency = (int64_t) millisec_to_nanosec(
+            cct->_conf->bluestore_codel_max_target_latency);
+    initial_bluestore_budget = cct->_conf->bluestore_codel_initial_budget_bytes;
+    min_bluestore_budget = cct->_conf->bluestore_codel_min_budget_bytes;
+    bluestore_budget_increment =
+            cct->_conf->bluestore_codel_budget_increment_bytes;
+    regression_history_size =
+            cct->_conf->bluestore_codel_regression_history_size;
+
+    bluestore_budget = initial_bluestore_budget;
+    min_bluestore_budget = initial_bluestore_budget;
+    max_queue_length = min_bluestore_budget;
+    fast_interval = initial_fast_interval;
+    target_latency = initial_target_latency;
+    min_latency = INITIAL_LATENCY_VALUE;
+    slow_interval_registered_bytes = 0;
+    regression_throughput_history.clear();
+    regression_target_latency_history.clear();
+    slow_interval_start = ceph::mono_clock::zero();
+  }
+
+  {
+    std::lock_guard l1{fast_timer_lock};
+    fast_timer.cancel_all_events();
+    fast_timer.init();
+    _fast_interval_process();
+  }
+  {
+    std::lock_guard l2{slow_timer_lock};
+    slow_timer.cancel_all_events();
+    slow_timer.init();
+    _slow_interval_process();
+  }
+}
+
+void BlueStoreSlowFastCoDel::reset_bluestore_budget() {
+  if (activated) {
```

Comment:
don't we need a lock here?

Reply:
That's right. I added the lock guard at line 120.